### ISSUE TITLE
Support AS[number] entries to block client requests from entire hosting servers

### ISSUE THREAD

**Issue by DandelionSprout** [AdguardTeam/AdGuardHome#1487]
### Prerequisites

Please answer the following questions for yourself before submitting an issue. **YOU MAY DELETE THE PREREQUISITES SECTION.**

- [x] I am running the latest version (v0.101.0)
- [x] I checked the documentation and found no answer
- [ ] I checked to make sure that this issue has not already been filed

### Problem Description

As per https://github.com/AdguardTeam/AdGuardHome/issues/1032#issuecomment-600063960.

In 'DNS settings', I request it could become possible to block requests from entire webhost servers at once through their official(?) AS numbers. For instance, the entirety ot `CARInet` could be blocked with *AS10439*, and the entirety of `Dedipath` (incl. their many subcontractors) could be blocked with *AS35913*. This could dramatically reduce the amount of effort and/or time that'd be needed to maintain anti-'Port scanner' client rejection lists.

There is however one big problem: It's not entirely clear to me which website has the best and most up-to-date info on which AS-number-using webhost that is using which IPs. https://ipinfo.io/ has the database that is the most easily accessible by laymen, at least when it comes to non-raw formats; but sometimes isn't entirely up to date on new IP address acquisitions.

However, depending on where AdGuard Home gets their hosting server names for the "Most common clients" statistic from (and whether they get that info on the cheap or not), tying those hosting server names to AS numbers that can be blocked, would likely be the best *and* easiest solution, **but** it'd currently fail to notice, process or block around a third of incoming requests, especially those shown in "Most common clients" as *Mobile end user NAT pool*

### Proposed Solution

In DNS Settings → Disallowed Clients, entries starting with `AS` and followed by digits, will be accepted and used to block requests from entire webhost servers at once.

### Alternatives Considered

None that I'm aware of.

### Additional Information

Many IP CIDR info sites, such as IPinfo, fail to properly compress the CIDRs listed for each webhost, leading to partially redundant CIDRs like `123.45.66.0/23` and `123.45.67.0/24` being listed for the same company at once. This is also something to consider when looking for the best IP address information source.

**Comment by HopHop20** [comment]
I do like this idea. I don't know the best way to implement this feature but I assume would be the use of an API like...

`https://bgpview.docs.apiary.io/#`

This could be done local on the users machine but making an API call to check ASN for each query could leave some slow resolutions on DNS requests and potential abuse of APIs.

Maybe one solution would be...

Check cache for known ASNs and their associated prefixes. Then simply deny or accept the request based on users rule.

Otherwise if there's nothing  cached look up the resolved ip of the query here
`https://api.bgpview.io/ip/8.8.8.8` and retrieve the ASN. Next lookup the ASN here `https://api.bgpview.io/asn/15169/prefixes` and cache all prefixes. And again deny or accept based on users rules.

In theory, if the user was only resolving domains that happen to fall under just one ASN, only a total of two API requests would be sent to `api.bgpview.io`

**Comment by ameshkov** [comment]
We can simply use ripe API directly:
https://stat.ripe.net/data/announced-prefixes/data.json?resource=AS3333

The lookup should be done periodically for all AS numbers from the blocklist/allowlist, and the result should be cached.

**Comment by DandelionSprout** [comment]
The following is more of a fun little thing for end-users rather than anything that'd help fix this issue outright, but after months of googling I've found possibly the one and only CIDR online compressor tool known to exist: https://tehnoblog.org/ip-tools/ip-address-aggregator/

With it, my efforts to block large repeat-offender hosting servers based on info from *IPinfo*, takes 5min at most, instead of 40~70min., which was particularly noticeable for `M247`.

**Comment by emlimap** [comment]
Another option could be is Maxmind's GeoLite2 ASN database. The database is updated monthly and offered as a downloadable file(mmdb) so you can run unlimited lookups locally without hitting any limits. The only downside is that they require a license key to download which is free for end users. End user would need to register at maxmind's website and get a license key. The procedure https://matomo.org/faq/how-to/how-do-i-get-a-license-key-for-the-maxmind-geolocation-database/

If AGH is to ask users to provide license key, then AGH can download as well as periodically update the database by itself. You just have to pass API key as a parameter to download URL https://dev.maxmind.com/geoip/geoip-direct-downloads/

They do have a commercial license which lets you re-distribute the file but that costs $252/year for IP 2 country and $456 for IP 2 City. No mention of IP 2 ASN.  https://www.maxmind.com/en/geolite2-commercial-redistribution

Not sure if they have any exceptions for open source projects as Flagfox(https://flagfox.wordpress.com/) a Firefox addon embeds their database  and ships with the addon for years. Even well after the license key was a requirement which happened in Jan 2020 (https://blog.maxmind.com/2019/12/18/significant-changes-to-accessing-and-using-geolite2-databases/)

Arch Linux (https://archlinux.org/packages/extra/any/geoip-database/) seems to getting it from this website which offers the legacy version https://mailfud.org/geoip-legacy

This is the info you get for an IP lookup
```
$ mmdblookup -f /usr/share/GeoIP/GeoLite2-ASN.mmdb -i 8.8.8.8

  {
    "autonomous_system_number":
      15169 <uint32>
    "autonomous_system_organization":
      "GOOGLE" <utf8_string>
  }
```

Potential Go libraries that could help with reading mmdb

https://github.com/oschwald/geoip2-golang
https://github.com/oschwald/maxminddb-golang
https://github.com/IncSW/geoip2


Also came across this one which seems to offer it in TSV file format and is updated hourly. Offered under Public Domain Dedication and License version v1.0 https://iptoasn.com/

Also, this one offers it as download without any registration under CC 4.0 license https://db-ip.com/db/download/ip-to-asn-lite
The details provided for an IP is similar to that of Maxmind's offering
```
$ mmdblookup -f dbip-asn-lite-2021-03.mmdb -i 8.8.8.8

  {
    "autonomous_system_number":
      15169 <uint32>
    "autonomous_system_organization":
      "Google LLC" <utf8_string>
  }
```

**Comment by ameshkov** [comment]
> Another option could be is Maxmind's GeoLite2 ASN database

Yeah, this is the easiest option and I should've mentioned it here right away.

**Comment by ainar-g** [comment]
We'll have to postpone this until the refactoring of the API is done.  This will require a lot of changes to implement properly.

**Comment by MotazHakim** [comment]
any update regarding this feature
blocking ASN is needed by me 
because i'm seeing alof of Cisco IPs using my DNS, 
like this CIDR
```
173.36.0.0/16
173.38.0.0/16
173.39.0.0/16
```
i had to add them like that instead of only blocking : AS109

since you added blocking by CIDRs, IP addresses, or ClientIDs, please Add ASN

thanks

**Comment by Barsovski** [comment]
#6715 closed as duplicate of this issue) 
But there almost four years have passed since the creation if this. 
Any news?

**Comment by PotatoPTT** [comment]
Temporary solution: copy and paste your country ip from this repo [country-ip-blocks](https://github.com/herrbischoff/country-ip-blocks) in allowlist 
I did this about 2 years ago and still working fine in all ip's that I've used in Brazil

**Comment by Barsovski** [comment]
> Temporary solution: copy and paste your country ip from this repo [country-ip-blocks](https://github.com/herrbischoff/country-ip-blocks) in allowlist I did this about 2 years ago and still working fine in all ip's that I've used in Brazil

What about blocking IP addresses using iptables? Tried this, but iptables doesn't seem to work for AGH requests(
